spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/source/SparkTable.java

```diff
@@ -186,22 +184,37 @@ public ScanBuilder newScanBuilder(CaseInsensitiveStringMap options) {
       icebergTable.refresh();
     }
 
-    SparkScanBuilder scanBuilder = new SparkScanBuilder(sparkSession(), icebergTable, options);
-
-    if (requestedSchema != null) {
-      scanBuilder.pruneColumns(requestedSchema);
-    }
-
-    return scanBuilder;
+    // If the table is loaded using the Spark DataFrame API, and option("snapshot-id", <snapshot_id>)
+    // or option("as-of-timestamp", <timestamp>)  is applied to the DataFrameReader, SparkTable will be
+    // constructed with a non-null snapshotId. Subsequently SparkTable#newScanBuilder will be called
+    // with the options, which will include "snapshot-id" or "as-of-timestamp".
+    // On the other hand, if the table is loaded using SQL, with the table suffixed with a snapshot
+    // or timestamp selector, then SparkTable will be constructed with a non-null snapshotId, but
+    // SparkTable#newScanBuilder will be called without the "snapshot-id" or "as-of-timestamp" option.
+    // We therefore add a "snapshot-id" option here in this latter case.
+    CaseInsensitiveStringMap scanOptions =
```

Comment:
I'm not sure this is worth the complexity. Why not just always add the snapshot ID?

Reply:
With the update to `addSnapshotId` below, this worked fine with tests:

```java
CaseInsensitiveStringMap scanOptions = addSnapshotId(options, snapshotId);
```

It didn't need the null check because that's done inside `addSnapshotId`.

Reply:
Let me try it out.
I had run into a problem with the original addSnapshot function and always calling it. After I analysed what was happening, I wrote that comment to remind myself. I therefore called addSnapshot only when strictly necessary.

Reply:
Ok, I see that you changed
```
      Preconditions.checkArgument(snapshotIdFromOptions == null,
          "Cannot override snapshot ID more than once: %s", snapshotIdFromOptions);
```
to
```
      Preconditions.checkArgument(snapshotIdFromOptions == null || snapshotId.toString().equals(snapshotIdFromOptions),
          "Cannot override snapshot ID more than once: %s", snapshotIdFromOptions);
```
in `addSnapshotId`.
With the old version, you should only call `addSnapshotId` if the `options` did not already have `snapshot-id` or `as-of-timestamp`.